NPE when shutting down JPA after a failed initialization

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

During model test execution, shutdown after a timed out test / failure shows a NullPointerException in https://github.com/keycloak/keycloak/actions/runs/5244964281/jobs/9472123944?pr=20430

```
023-06-12T15:12:44.0412060Z 15:12:43,937 DEBUG [org.keycloak.testsuite.model.KeycloakModelParameters] (pool-10-thread-3) Closing KeycloakSessionFactory 11 (from pool-10-thread-3 thread)
2023-06-12T15:12:44.0413186Z 15:12:43,938 ERROR [org.keycloak.testsuite.model.KeycloakModelParameters] (Time-limited test) Execution didn't complete
2023-06-12T15:12:44.0414016Z java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "java.lang.Runnable.run()" because "this.onClose" is null
2023-06-12T15:12:44.0414856Z    at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
2023-06-12T15:12:44.0415403Z    at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:205)
2023-06-12T15:12:44.0416442Z    at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactories(KeycloakModelTest.java:430)
2023-06-12T15:12:44.0417763Z    at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.testPersistenceMultipleNodesClientSessionsAtRandomNode(OfflineSessionPersistenceTest.java:249)
2023-06-12T15:12:44.0418968Z    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-06-12T15:12:44.0419750Z    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-06-12T15:12:44.0420605Z    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-06-12T15:12:44.0421323Z    at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-06-12T15:12:44.0421955Z    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2023-06-12T15:12:44.0422706Z    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2023-06-12T15:12:44.0423487Z    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2023-06-12T15:12:44.0424261Z    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2023-06-12T15:12:44.0425042Z    at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
2023-06-12T15:12:44.0425948Z    at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
2023-06-12T15:12:44.0426637Z    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-06-12T15:12:44.0427238Z    at java.base/java.lang.Thread.run(Thread.java:833)
2023-06-12T15:12:44.0427916Z Caused by: java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "java.lang.Runnable.run()" because "this.onClose" is null
2023-06-12T15:12:44.0428789Z    at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:501)
2023-06-12T15:12:44.0429656Z    at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$6(KeycloakModelTest.java:401)
2023-06-12T15:12:44.0430482Z    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-06-12T15:12:44.0431164Z    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2023-06-12T15:12:44.0431899Z    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2023-06-12T15:12:44.0432415Z    ... 1 more
2023-06-12T15:12:44.0432960Z Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Runnable.run()" because "this.onClose" is null
2023-06-12T15:12:44.0433732Z    at org.keycloak.models.jpa.JpaRealmProviderFactory.close(JpaRealmProviderFactory.java:70)
2023-06-12T15:12:44.0434648Z    at org.keycloak.services.DefaultKeycloakSessionFactory.close(DefaultKeycloakSessionFactory.java:403)
2023-06-12T15:12:44.0435577Z    at org.keycloak.testsuite.model.KeycloakModelTest.createKeycloakSessionFactory(KeycloakModelTest.java:352)
2023-06-12T15:12:44.0436586Z    at org.keycloak.testsuite.model.KeycloakModelTest.reinitializeKeycloakSessionFactory(KeycloakModelTest.java:362)
2023-06-12T15:12:44.0437790Z    at org.keycloak.testsuite.model.session.OfflineSessionPersistenceTest.lambda$testPersistenceMultipleNodesClientSessionsAtRandomNode$27(OfflineSessionPersistenceTest.java:277)
2023-06-12T15:12:44.0438883Z    at org.keycloak.testsuite.model.KeycloakModelTest.lambda$inIndependentFactories$5(KeycloakModelTest.java:408)
2023-06-12T15:12:44.0439768Z    at org.keycloak.testsuite.model.KeycloakModelTest.inIndependentFactory(KeycloakModelTest.java:499)
2023-06-12T15:12:44.0440360Z    ... 5 more
```


### Version

main

### Expected behavior

When JPA isn't fully initialized on shutdown, it shouldn't throw an exception.

### Actual behavior

Throws NPE

### How to Reproduce?

Random occurrence in the test after a test timed out ("[org.keycloak.testsuite.model.KeycloakModelParameters] (Time-limited test) Execution didn't complete")

### Anything else?

I'll prepare a PR.

--- a/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProviderFactory.java
+++ b/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProviderFactory.java
@@ -67,7 +67,9 @@ public class JpaRealmProviderFactory implements RealmProviderFactory, ProviderEv
 
     @Override
     public void close() {
-        onClose.run();
+        if (onClose != null) {
+            onClose.run();
+        }
     }
 
     @Override

--- a/model/legacy-private/src/main/java/org/keycloak/storage/datastore/LegacyDatastoreProviderFactory.java
+++ b/model/legacy-private/src/main/java/org/keycloak/storage/datastore/LegacyDatastoreProviderFactory.java
@@ -66,7 +66,9 @@ public class LegacyDatastoreProviderFactory implements DatastoreProviderFactory,
 
     @Override
     public void close() {
-        onClose.run();
+        if (onClose != null) {
+            onClose.run();
+        }
     }
 
     @Override

